File: backend/src/models/topic.dart
class Topic {
  final String id;
  final String subjectId;
  final String name;
  final bool completed;
  final DateTime? completedAt;

  Topic({
    required this.id,
    required this.subjectId,
    required this.name,
    required this.completed,
    this.completedAt,
  });

  factory Topic.fromJson(Map<String, dynamic> json) {
    return Topic(
      id: json['_id']?.toString() ?? '',
      subjectId: json['subjectId']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      completed: json['completed'] == true,
      completedAt: json['completedAt'] != null
          ? DateTime.tryParse(
              json['completedAt'].toString(),
            )
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'subjectId': subjectId,
      'name': name,
      'completed': completed,
      'completedAt': completedAt?.toIso8601String(),
    };
  }
}

File: backend/src/models/Topic.js
const mongoose = require("mongoose");

const topicSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
      index: true,
    },

    subjectId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Subject",
      required: true,
      index: true,
    },

    unitId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Unit",
      required: true,
      index: true,
    },

    name: {
      type: String,
      required: true,
      trim: true,
    },

    completed: {
      type: Boolean,
      default: false,
    },

    completedAt: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

/*
 * A topic name must be unique inside
 * the same user's subject + unit.
 */
topicSchema.index(
  {
    userId: 1,
    subjectId: 1,
    unitId: 1,
    name: 1,
  },
  {
    unique: true,
  }
);

module.exports =
  mongoose.model(
    "Topic",
    topicSchema
  );